lib/modules.nix

```diff
@@ -233,9 +233,7 @@ rec {
       if opt.options ? default && res ? default ||
          opt.options ? example && res ? example ||
          opt.options ? description && res ? description ||
-         opt.options ? apply && res ? apply ||
-         # Accept to merge options which have identical types.
-         opt.options ? type && res ? type && opt.options.type.name != res.type.name
```

Comment:
The goal of this check was to prevent the extension of additional option declarations with a type which is not identical.  For example, submodule option has a documentation string (`name`), thus we can extend them with additional submodules.  Another example, is with 2 identical types, such as `uniq string` can be merged with each others.

So this is technically a regression.

The generic question is more subtle than that, which is, is the merge procedure between these 2 types commutative?  But honestly, we don't want to answer this one, because in practice this does not hold true, the reason being that the order of the merge is unspecified.

With the addition of `typeMerge` function, I would recommend to drop the use of these documentation strings (`name`), in favor of easily comparable types which can be used to uniquely identify a type, and also in a way to carry over the values out-side the comparison set.

So, one potential option would be to replace the `inhabitant` attribute, by a `functor` attribute, which can be compared and merged.

```nix
{
  enum = values: mkOptionType {
    functor = { type = enum; wrapped = []; payload = [ values ]; };
  };

  loaOf = elemType: mkOptionType {
    functor = { type = loaOf; wrapped = [ elemType ]; payload = []; };
  };
}
```

Thus merging option declarations will recursively check for if the functions are identical, and if we can zip the arguments, otherwise we should throw.

What do you think?